### PR TITLE
[TEVA-1186] refactor filter group into a view component

### PR DESCRIPTION
**Pull request by alexbowen**
https://dfedigital.atlassian.net/browse/TEVA-1186

- refactors filters to view component
- all filters look/behave exactly as before

### CODE REVIEW

File: spec/components/shared/filters_component_spec.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      expect(inline_component.css('.moj-filter__content').to_html).to be_blank
+
+      expect(
+        inline_component.css('.moj-filter__content .govuk-heading-s').to_html
+      ).to be_blank
+
+      expect(inline_component.css('.filters__heading-applied').to_html).to be_blank
+    end
+  end
+end
```
Comment (alexbowen) [inline spec/components/shared/filters_component_spec.rb:52]: some help needed here as the formbuilder object doesnt seem to do what i expect and hence tests are failing

File: spec/components/shared/filters_component_spec.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
+require 'rails_helper'
+
+RSpec.describe Shared::FiltersComponent, type: :component do
+  let(:f) { VacancyAlgoliaSearchForm.new() }
```
Comment (cpjmcquillan) [inline spec/components/shared/filters_component_spec.rb:4]: `f` should be an instance of `GOVUKDesignSystemFormBuilder::FormBuilder`

Comment (alexbowen) [reply spec/components/shared/filters_component_spec.rb:4]: ah ok thanks, give it ago

Comment (cpjmcquillan) [reply spec/components/shared/filters_component_spec.rb:4]: Maybe something like this?
```ruby
GOVUKDesignSystemFormBuilder::FormBuilder.new(:search_form, VacancyAlgoliaSearchForm.new, rendered_component, {})
```

Comment (cpjmcquillan) [reply spec/components/shared/filters_component_spec.rb:4]: This is what `f` needs to refer to https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-c-new

Comment (alexbowen) [reply spec/components/shared/filters_component_spec.rb]: i stubbed the formbuilder object instead as i couldnt get your suggestion to work?!

File: app/components/shared/filters_component.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+class Shared::FiltersComponent < ViewComponent::Base
```
Comment (cpjmcquillan) [inline app/components/shared/filters_component.rb:1]: can we use `attr_accessor :items, ...` here please?

File: app/components/shared/filters_component.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  def initialize(items:, filters:, form:, options:)
+    @filters = filters
```
Comment (cpjmcquillan) [inline app/components/shared/filters_component.rb:3]: Can these be in the same order?

File: app/components/shared/filters_component/filters_component.html.haml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    - if display_remove_buttons
+      .moj-filter__content
+        .moj-filter__selected
+          .filters__section-heading{ class: 'govuk-!-margin-bottom-2' }
+            .govuk-body{ class: 'govuk-!-margin-bottom-0' }
+              = t('shared.filter_group.current_selected_filters')
+            %button.govuk-link.govuk-link--no-visited-state{ id: 'clear-filters-button' }
+              = t('shared.filter_group.clear_all_filters')
+
+          - @items.each do |group|
```
Comment (cpjmcquillan) [inline app/components/shared/filters_component/filters_component.html.haml:25]: Would it make more sense to name these the same way? E.g `groups.each do |group|` or `items.each do |item|`

File: spec/components/shared/filters_component_spec.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+RSpec.describe Shared::FiltersComponent, type: :component do
+  class Form
```
Comment (cpjmcquillan) [inline spec/components/shared/filters_component_spec.rb:4]:
```ruby
let(:f) { instance_double(GOVUKDesignSystemFormBuilder::FormBuilder) }

before do
  allow(f).to receive(:govuk_submit)
  allow(f).to receive(:govuk_collection_check_boxes)
end
```

File: spec/components/shared/filters_component_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    def govuk_submit(arg1, arg2); end
+
+    def govuk_collection_check_boxes(arg1, arg2, arg3, arg4, arg5, arg6); end
+  end
+
+  let(:f) { Form.new }
+
+  context 'when there are selected filters' do
+    subject do
+      described_class.new(form: f, filters: { total_count: 2 }, items: [{ title: 'Group 1', key: 'group_1', options: [%w[option_1 OPTION_1]], selected: %w[option_1], value_method: :first, selected_method: :last }, { title: 'Group 2', key: 'group_2', options: [%w[option_1 OPTION_1]], selected: %w[option_1], value_method: :first, selected_method: :last }], options: { remove_buttons: true, close_all: true })
```
Comment (cpjmcquillan) [inline spec/components/shared/filters_component_spec.rb:14]: Can we break this over several lines so it fits in the line limit please? Rubocop should pick this up 🤔

File: spec/components/shared/filters_component_spec.rb (hunk 6 above)
Comment (alexbowen) [reply spec/components/shared/filters_component_spec.rb:4]: 👍

File: spec/components/shared/filters_component_spec.rb (hunk 7 above)
Comment (alexbowen) [reply spec/components/shared/filters_component_spec.rb:14]: no rubocop didnt obviously? but we are updating that config soon arent we?

File: app/components/shared/filters_component.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
+class Shared::FiltersComponent < ViewComponent::Base
+  attr_accessor :filters, :form, :items, :options
+
+  def initialize(filters:, form:, items:, options:)
+    @filters = filters
+    @f = form
```
Comment (cpjmcquillan) [inline app/components/shared/filters_component.rb:6]: Could this be `@form = form` please?

File: app/components/shared/filters_component/filters_component.html.haml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+.filters
+  .filters__heading.moj-filter
+    %div{ class: "#{@options[:close_all] ? 'filters__heading-container' : 'moj-filter__header'}" }
+      .filters__section-heading{ class: mobile_modifier('filters__heading') }
+        .moj-filter__heading-title
+          %h2.govuk-heading-s{ class: mobile_modifier('filters__heading-title'), id: "#{@options[:mobile_variant] ? 'mobile-filters-button' : ''}" }
+            = @filters[:title]
+          %span.govuk-body.filters__heading-applied= applied_text
+          %button.govuk-link.govuk-link--no-visited-state.filters--show-mobile-open{ id: 'return-to-results' }
+            = 'Return to results'
```
Comment (cpjmcquillan) [inline app/components/shared/filters_component/filters_component.html.haml:10]: I18n!

File: app/components/hiring_staff/vacancies_component.html.haml [hunk 10]
```diff
@@ -2,7 +2,7 @@
   - if @organisation.is_a?(SchoolGroup)
     .govuk-grid-column-one-third
       = form_for @filters_form, url: organisation_managed_organisations_path(anchor: 'vacancies-hits'), html: { method: "patch", data: { 'auto-submit': true } } do |f|
-        = render 'shared/filter_group', f: f, filters: { title: t('jobs.filters.job_filters'), remove_buttons: true, total_count: @filters[:managed_school_ids]&.count, items: [{ title: 'Locations', key: 'locations', search: true, scroll: true, attribute: :managed_school_ids, selected: @filters[:managed_school_ids], options: @organisation_options, value_method: :id, text_method: :label, selected_method: :name, small: true }] }
+        = render partial: 'components/shared/filters_component', locals: { form: f, options: { remove_buttons: true }, filters: { total_count: @filters[:managed_school_ids]&.count, title: t('jobs.filters.job_filters') }, items: [{ title: 'Locations', key: 'locations', search: true, scroll: true, attribute: :managed_school_ids, selected: @filters[:managed_school_ids], options: @organisation_options, value_method: :id, text_method: :label, selected_method: :name, small: true }] }
```
Comment (cpjmcquillan) [inline app/components/hiring_staff/vacancies_component.html.haml:5]: I think render with 'partial' and 'locals' is the old style

File: app/views/vacancies/_filters.html.haml [hunk 11]
```diff
@@ -11,4 +11,4 @@
...
+    = render partial: 'components/shared/filters_component', locals: { form: f, options: { remove_buttons: true, mobile_variant: true, close_all: true }, filters: { total_count: @jobs_search_form.total_filters, title: 'Filter results' }, items: [{ title: 'Job role', key: 'job_roles', search: false, scroll: false, attribute: :job_roles, selected: @jobs_search_form.job_roles, options: @jobs_search_form.job_role_options, value_method: :first, text_method: :last, selected_method: :last, small: true }, { title: 'Education phase', key: 'education_phase', search: false, scroll: false, attribute: :phases, selected: @jobs_search_form.phases, options: @jobs_search_form.phase_options, value_method: :first, text_method: :last, selected_method: :last, small: true }, { title: 'Working pattern', key: 'working_patterns', search: false, scroll: false, attribute: :working_patterns, selected: @jobs_search_form.working_patterns, options: @jobs_search_form.working_pattern_options, value_method: :first, text_method: :last, selected_method: :last, small: true }] }
```
Comment (cpjmcquillan) [inline app/views/vacancies/_filters.html.haml:14]: Same as above

File: app/components/shared/filters_component/filters_component.html.haml [hunk 12]
```diff
@@ -0,0 +1,61 @@
+.filters
+  .filters__heading.moj-filter
+    %div{ class: "#{@options[:close_all] ? 'filters__heading-container' : 'moj-filter__header'}" }
```
Comment (cpjmcquillan) [inline app/components/shared/filters_component/filters_component.html.haml:3]: Now we're using `attr_accessor` in the component, we should be able to access all of the variables (don't need to access the instance variable)

**Review by cpjmcquillan — APPROVED** [review]

**Pull request merged**